### ISSUE TITLE
knockout and tableSorter with computed array - problems with pager and search box

### ISSUE THREAD

**Issue by biocar2001** [Mottie/tablesorter#915]
Hi,

We have a table which one show information about users. When We are using this control with knockout and we iterate the table with a observable Array everything is fine with the help of a handler developed for us:

```js
ko.bindingHandlers.sortTableComputed = {
	init: function(element, valueAccessor) {
	    var iden=$(element).attr("id");
	    if($(element).hasClass('tablesorter')){
	        $(element).trigger('updateAll');
	    }else{
	        $(element).tablesorter({
	            widgetOptions:{filter_external : '#'+iden+"_search"}
	        }).tablesorterPager({container: $('#'+iden+"_pager")});
	    }
	},
	update: function(element, valueAccessor) {
	  ko.utils.unwrapObservable(valueAccessor()); 
	  setTimeout(function() {
	    $(element).trigger("update");
	    $(element).trigger('pagerUpdate');
	  }, 0);
	}
};
```

and in the template we do binding in this way:

```html
<div id="tablaAccionesKO_pager" class="col-lg-10 col-md-10 form-horizontal tablesorter-pager">
    <button type="button" class="btn btn-default first disabled" tabindex="0" aria-disabled="true"><i class="glyphicon glyphicon-step-backward"></i></button>
    <button type="button" class="btn btn-default prev disabled" tabindex="0" aria-disabled="true"><i class="glyphicon glyphicon-backward"></i></button>
    <span class="pagedisplay">1 - 10 / 50 (50)</span> <!-- this can be any element, including an input -->
    <button type="button" class="btn btn-default next" tabindex="0" aria-disabled="false"><i class="glyphicon glyphicon-forward"></i></button>
    <button type="button" class="btn btn-default last" tabindex="0" aria-disabled="false"><i class="glyphicon glyphicon-step-forward"></i></button>
    <select class="pagesize" title="Select page size" aria-disabled="false">
        <option value="5">5</option>
        <option selected="selected" value="10">10</option>
        <option value="25">25</option>
        <option value="50">50</option>
        <option value="100">100</option>
        <option value="200">200</option>
        <option value="9999">Todo</option>
    </select>
    <select class="pagenum" title="Select page number" aria-disabled="false"><option>1</option><option>2</option><option>3</option><option>4</option><option>5</option></select>
</div>
<div class="input-group col-lg-2 col-md-2 tablesorter-pager">
  <span class="input-group-addon"><span class="glyphicon glyphicon-search"></span></span>
  <input id="tablaAccionesKO_search" type="search" class="form-control" placeholder="buscar" data-column="all"></input>
</div>
<table class="table table-striped" id="tablaAccionesKO" data-bind="sortTableComputed">
    <thead>
        <tr>
            <th></th>
            <th class='columnaId'><span data-toggle='tooltip'>Id</span></th>
            <th>Tipo</th>
            <th><span class='glyphicon glyphicon-share ayuda' title='Proyectos Relacionados' data-bind="tooltip"></span></th>
            <th><span class='glyphicon glyphicon-list ayuda' title='Categoria' data-bind="tooltip"></span></th>
            <th>Responsables</th>
            <th><span class='ayuda' title='Horas incurridas' data-bind="tooltip"><span class='glyphicon glyphicon-time'><span class='glyphicon glyphicon-wrench'></th>
            <th>T&iacute;tulo</th>
            <th class='columnaFecha'>Fin.Plan</th>
            <th class='columnaFecha'>Fin.Real</th>
            <th>%</th>
            <th>Estado</th>
            <th class='botonesFijos'>Botonesssss</th>
        </tr>
    </thead>
    <tbody data-bind="foreach: listaAccionesFiltrada">
```

The problem happen when our MyListObservable is a computed value, which one is recalculated in function of other check observable. In this case the behaviour of the control is strange and always add elements to the table but when the observable array remove information this one dont update the index of the table and theirs values. I have tried without success:

1º RemoveAll the values of the array before reload the array.
2º Force the updateAll the table just after of reload of computed value $(element).trigger('updateAll').
3º Get the value of the computed and assign his value to other observable array and iterate with this new array.

I know here is the repository of TableSorter and not of knockout but obviously the problem has relation with you and maybe other users had the same problem.
We have tried refresh the pagination and search box, but both dont work fine neither. I wrote the same question in stackoverflow and the repository o knockout Team by anyone knows what happen or what is the problem.

https://github.com/knockout/knockout/issues/1791
http://stackoverflow.com/questions/30218970/table-with-tablesorter-in-knockout-with-computed-value-for-iteration

Any idea?
Thanks in advance.

**Comment by Mottie** [comment]
Hi @biocar2001!

I don't use knockout, so I'm not very familiar with how it works. After reading the contents of the links you shared, I think I might know what is needed to fix this issue.

You didn't state this about knockout, but from looking at the Plunker in the answer on Stackoverflow, I see that the build widget is needed to process & build the table from the knockout array. The build widget was originally written to only work on initialization; but I think you can use it instead of calling "update" or "updateAll". *I haven't tried this method before*, but the following code looks like it should work:

```js
init: function(element, valueAccessor) {
    var iden=$(element).attr("id");
    if($(element).hasClass('tablesorter')){
        $.tablesorter.buildTable( element, element.config );
    }else{
        $(element).tablesorter({
            widgetOptions:{filter_external : '#'+iden+"_search"}
        }).tablesorterPager({container: $('#'+iden+"_pager")});
    }
},
```

This code assumes that `element` is the `table`. `element` could also be the wrapper, but then it would append a new table.

**Comment by biocar2001** [comment]
Thanks by your answer but if I try it, assuming the table is element of course:

ko.bindingHandlers.sortTableComputed = {
	init: function(element, valueAccessor) {
	    var iden=$(element).attr("id");
	    if($(element).hasClass('tablesorter')){
	        $.tablesorter.buildTable( element, element.config );
	        $(element).trigger('update');
	    }else{
	        $(element).tablesorter({
	            widgetOptions:{filter_external : '#'+iden+"_search"}
	        }).tablesorterPager({container: $('#'+iden+"_pager")});
	    }
	},

	update: function(element, valueAccessor) {
	  ko.utils.unwrapObservable(valueAccessor()); // adds dependency, so when associated observable value is changed "update" method will be triggered
	  setTimeout(function() {
	    $(element).trigger("update");
	  }, 0);
	}
};

When the computed array change, the table and the pager of the table never is refresh, I suppose this one is not the best channel for these questions becasue you arent master of knockout but I dont know where can I ask......I will continue trying this solution otherwise We will try use othe UI component.

Thanks.

**Comment by Mottie** [comment]
I wonder, does the `init` get called a second time or just the `update` function?

Maybe try moving that call to the `buildTable` function inside the `update`:

```js
ko.bindingHandlers.sortTableComputed = {
    init: function(element, valueAccessor) {
        var iden=$(element).attr("id");
        if($(element).hasClass('tablesorter')){
            $(element).trigger('update');
        }else{
            $(element).tablesorter({
                widgetOptions:{filter_external : '#'+iden+"_search"}
            }).tablesorterPager({container: $('#'+iden+"_pager")});
        }
    },

    update: function(element, valueAccessor) {
      ko.utils.unwrapObservable(valueAccessor()); // adds dependency, so when associated observable value is changed "update" method will be triggered
      setTimeout(function() {
        $.tablesorter.buildTable( element, element.config );
        $(element).trigger("update");
      }, 0);
    }
};
```

**Comment by biocar2001** [comment]
Thanks again but the second solution doesnt work neither............. I am desperate. Now I am trying to do other solution with jquery, so I would like refresh the pager widget(Pagination and index) after refresh some values in the array wich one we use for iterate in the tablesorter. I have tried everything:
console.log(Object.getOwnPropertyNames($('#tablaCatalogo')[0].config));

//Nothing        
var sorting = $('#tablaCatalogo')[0].config.sortList;
        $("#tablaCatalogo").unbind('filterStart.pager')
            .bind('filterStart.pager', function(e, filters) {
                this.config.pager.currentFilters = filters;
            });
        $("#tablaCatalogo").
            trigger("updateAll")
            .trigger("updateCache");
            .trigger("applyWidgets");
            .trigger('sorton', [sorting]);
        
        $("#tablaCatalogo").trigger("updateAll");
        $("#tablaCatalogo").trigger("appendCache");

I don find solution for my problem, now I only need the pager widget(Pagination,index and search) refresh after I delete or I hide some registers in the table, do you understand? could you help me? I suppose this thing shoud be so easy I can not get the solution.

Thanks a lot.

**Comment by Mottie** [comment]
Try the [`pagerUpdate` method](http://mottie.github.io/tablesorter/docs/#pagerupdate).

**Comment by biocar2001** [comment]
Ok that`s great but could you tell me how can I extend this method? Our case We are hidding registers and We are not deleting the register, for this reason I suppose the method arent doing what I want, this method is showing all the registers the array contains, but in our case we want show only the registers wich one are visible in the table. would it be possible?

Thanks.

**Comment by Mottie** [comment]
I'm not sure I understand... you're saying that `pagerUpdate` shows all rows?

Have you tried the [`pageAndSize` method](http://mottie.github.io/tablesorter/docs/#pageandsize)?

**Comment by biocar2001** [comment]
Yes, pagerUpdate shows all rows but pageAndSize method Dont, but I dont understadn how this method can help me..........

**Comment by biocar2001** [comment]
I dont kno, maybe the proble is how we inicializate the table:
function inicializaTableSorter(){
  $.extend($.tablesorter.themes.bootstrap, {
    // these classes are added to the table. To see other table classes available,
    // look here: http://twitter.github.com/bootstrap/base-css.html#tables
    table      : 'table',
    caption    : 'caption',
    header     : 'bootstrap-header', // give the header a gradient background
    footerRow  : '',
    footerCells: '',
    icons      : '', // add "icon-white" to make them white; this icon class is added to the <i> in the header
    sortNone   : '',//'bootstrap-icon-unsorted',
    sortAsc    : 'glyphicon glyphicon-chevron-up',     // includes classes for Bootstrap v2 & v3
    sortDesc   : 'glyphicon glyphicon-chevron-down', // includes classes for Bootstrap v2 & v3
    active     : '', // applied when column is sorted
    hover      : '', // use custom css here - bootstrap class may not override it
    filterRow  : '', // filter row class
    even       : '', // odd row zebra striping
    odd        : ''  // even row zebra striping
  });
  $.tablesorter.defaults.theme = "bootstrap";
  $.tablesorter.defaults.dateFormat = "ddmmyyyy";
  $.tablesorter.defaults.emptyTo = 'bottom';
  $.tablesorter.defaults.widthFixed = true;
  $.tablesorter.defaults.headerTemplate = '{content} {icon}';
  $.tablesorter.defaults.widgets = ["uitheme", "filter", "editable","staticRow"];
  $.tablesorter.defaults.sortList = [[0,1]];
  $.tablesorter.defaults.widgetOptions = {
    zebra : ["even", "odd"],
    filter_columnFilters: false,
    filter_saveFilters : true,
    filter_reset: '.reset'
  };
  $.tablesorterPager.defaults.output = '{startRow} - {endRow} / {filteredRows} ({totalRows})';
  $.tablesorterPager.defaults.size = 10;
  $.tablesorterPager.defaults.cssGoto = ".pagenum";
  $.tablesorterPager.defaults.removeRows = false;
  
}

**Comment by Mottie** [comment]
It's not that initialization function - it looks like it is only changing default settings.

Have you tried setting the `debug` option to `true` and seeing if the cache updates?

**Comment by biocar2001** [comment]
Hi,

First at all, I am sorry for my delay in the answer but I was working in others topics. Finally I used this option in knockout:

 /*Extends Pagination*/
	(function() {
	    'use strict';
	    ko.extenders.paging = function(target, pageSize) {
	        var _pageSize = ko.observable(pageSize || 10),
	            // default pageSize to 10
	            _currentPage = ko.observable(1); // default current page to 1
	        target.pageSize = ko.computed({
	            read: _pageSize,
	            write: function(newValue) {
	                if (newValue > 0) {
	                    _pageSize(newValue);
	                }
	                else {
	                    _pageSize(10);
	                }
	            }
	        });

	        target.currentPage = ko.computed({
	            read: _currentPage,
	            write: function(newValue) {
	                if (newValue > target.pageCount()) {
	                    _currentPage(target.pageCount());
	                }
	                else if (newValue <= 0) {
	                    _currentPage(1);
	                }
	                else {
	                    _currentPage(newValue);
	                }
	            }
	        });

	        target.pageCount = ko.computed(function() {
	            return Math.ceil(target().length / target.pageSize()) || 1;
	        });

	        target.currentPageData = ko.computed(function() {
	            var pageSize = _pageSize(),
	                pageIndex = _currentPage(),
	                startIndex = pageSize * (pageIndex - 1),
	                endIndex = pageSize * pageIndex;

	            return target().slice(startIndex, endIndex);
	        });

	        target.moveFirst = function() {
	            target.currentPage(1);
	        };
	        target.movePrevious = function() {
	            target.currentPage(target.currentPage() - 1);
	        };
	        target.moveNext = function() {
	            target.currentPage(target.currentPage() + 1);
	        };
	        target.moveLast = function() {
	            target.currentPage(target.pageCount());
	        };

	        return target;
	    };
	}());

With this extends,  you can appply this solution to a computedArray:
root.list = ko.computed(function() {...........},root).extend({paging: root.numPag()});

And in the template:
''' <div id="tableKO_pager" class="col-lg-10 col-md-10 form-horizontal tablesorter-pager">
                            <button type="button" class="btn btn-default first" tabindex="0"  data-bind="click: list.moveFirst"><i class="glyphicon glyphicon-step-backward"></i></button>
                            <button type="button" class="btn btn-default prev" tabindex="0" data-bind="click: list.movePrevious"><i class="glyphicon glyphicon-backward"></i></button>
                            
                            <button type="button" class="btn btn-default next" tabindex="0" aria-disabled="false" data-bind="click: list.moveNext"><i class="glyphicon glyphicon-forward"></i></button>
                            <button type="button" class="btn btn-default last" tabindex="0" aria-disabled="false" data-bind="click: list.moveLast"><i class="glyphicon glyphicon-step-forward"></i></button>
                            <select class="pagesize" title="Select page size" aria-disabled="false" data-bind="value: numPag, event: {change: changeNumPag(numPag())}">
                                <option selected="selected" value="5">5</option>
                                <option value="10">10</option>
                                <option value="25">25</option>
                                <option value="50">50</option>
                                <option value="100">100</option>
                                <option value="200">200</option>
                                <option value="9999">Todo</option>
                            </select>
                            Page Number:     
                            <input class="pagesize" type="number" data-bind="value: list.currentPage, attr: { 'max': listaAccionesFiltrada.pageCount }" min="1">
                                of <span data-bind="text: list.pageCount"></span>
                        </div>''' 
and this other in your iterate:
body data-bind='for each: list.currentPageData'

With this development you get a pagination with the styles of bootstrap-knockout but I lose the search control and ordenation by colum of your control...... I think these other options are impossible because in knockout when you are working with computed arrays, these arrays lose the features and methods of sort, although probably I could develop the search function, for now is enough for my requeriments.

Maybe with this source you can extend your control and allow use the control with computedArrays and knockout, I dont know....... anyway thanks a lot by your help.

Regards.

**Comment by Mottie** [comment]
Hey @biocar2001!

I was just looking at your code, and I think you can use the internal [pager variables](http://mottie.github.io/tablesorter/docs/#pager-variables) to accomplish the same thing.

Like I said, I don't know knockout, but you can access any of the internal pager variables from `table.config.pager`. From here you have access to:

* `pager.startRow` - one-based index of the first row visible in the pager.
* `pager.endRow` - one-based index of the last row visible in the pager.
* `pager.page` - currently selected page
* `pager.size` - currently selected page size.
* `pager.totalPages` - total pages determined by the current page size.
* `pager.totalRows` - total rows within the table, or returned by ajax.
* `pager.filteredPages` - filtered pages accessible by the pager. This equals `totalPages` if no filters are applied.
* `pager.filteredRows` - filtered rows accessible by pager. This equals `totalRows` if no filters are applied.
* `pager.$container` - pager outer wrapper (jQuery object)
* `pager.$goto` - pager page selector (jQuery object)
* `pager.$size` - pager size selector (jQuery object)

Then to set the page and/or size, use any of the available [pager methods](http://mottie.github.io/tablesorter/docs/#methods-pager).
